File: python-examples/11-error-handling/worker.py
"""
Polos Worker for the Error Handling example.

Run with:
    python worker.py
"""

import asyncio
import os

from dotenv import load_dotenv
from polos import PolosClient, Worker

from workflows import (
    retry_example,
    error_recovery,
    fallback_pattern,
    circuit_breaker,
    compensation_pattern,
)

load_dotenv()


async def main():
    """Main function to run the worker."""
    project_id = os.getenv("POLOS_PROJECT_ID")
    if not project_id:
        raise ValueError(
            "POLOS_PROJECT_ID environment variable is required. "
            "Get it from the Polos UI at http://localhost:5173/projects/settings"
        )

    client = PolosClient(
        project_id=project_id,
        api_url=os.getenv("POLOS_API_URL", "http://localhost:8080"),
    )

    worker = Worker(
        client=client,
        deployment_id="error-handling-examples",
        workflows=[
            retry_example,
            error_recovery,
            fallback_pattern,
            circuit_breaker,
            compensation_pattern,
        ],
    )

    print("Starting Error Handling Examples worker...")
    print(f"  Project ID: {project_id}")
    print(f"  Workflows: {[w.id for w in worker.workflows]}")
    print("  Press Ctrl+C to stop\n")

    try:
        await worker.run()
    except KeyboardInterrupt:
        print("\nShutting down worker...")
        await worker.shutdown()


if __name__ == "__main__":
    asyncio.run(main())

File: typescript-examples/11-error-handling/worker.ts
/**
 * Polos Worker for the Error Handling example.
 *
 * Run with:
 *   npx tsx worker.ts
 *
 * Environment variables:
 *   POLOS_PROJECT_ID - Your project ID (required)
 *   POLOS_API_URL - Orchestrator URL (default: http://localhost:8080)
 *   POLOS_API_KEY - API key for authentication (optional for local development)
 *   POLOS_DEPLOYMENT_ID - Deployment ID (default: error-handling-examples)
 */

import 'dotenv/config';
import { Worker } from '@polos/sdk';
import {
  retryExample,
  errorRecovery,
  fallbackPattern,
  circuitBreaker,
  compensationPattern,
} from './workflows.js';

async function main() {
  const projectId = process.env['POLOS_PROJECT_ID'];
  if (!projectId) {
    throw new Error(
      'POLOS_PROJECT_ID environment variable is required. ' +
        'Set it to your project ID (e.g., export POLOS_PROJECT_ID=my-project). ' +
        'You can get this from the output printed by `polos-server start` or from the UI page at ' +
        "http://localhost:5173/projects/settings (the ID will be below the project name 'default')",
    );
  }

  const apiUrl = process.env['POLOS_API_URL'] ?? 'http://localhost:8080';
  const apiKey = process.env['POLOS_API_KEY'] ?? '';
  const deploymentId = process.env['POLOS_DEPLOYMENT_ID'] ?? 'error-handling-examples';

  const worker = new Worker({
    apiUrl,
    apiKey,
    projectId,
    deploymentId,
    workflows: [retryExample, errorRecovery, fallbackPattern, circuitBreaker, compensationPattern],
  });

  console.log('Starting Error Handling Examples worker...');
  console.log(`  Project ID: ${projectId}`);
  console.log(`  Workflows: [${[retryExample, errorRecovery, fallbackPattern, circuitBreaker, compensationPattern].map((w) => w.id).join(', ')}]`);
  console.log('  Press Ctrl+C to stop\n');

  await worker.run();
}

main().catch(console.error);
